### PR TITLE
Lesion filtering using uncertainty -- DEV V1

### PR DESCRIPTION
**Pull request by charleygros**
### General idea:
To use the uncertainty information to perform a voxel-wise or a lesion-wise filtering in post-processing.
Based on [Nair et al. 2020 work](https://www.sciencedirect.com/science/article/pii/S1361841519300994#sec0010).

### Dev script:
The script `dev/filtering_lesion.py` computes the voxel-level perforamance ROC curves that compare performance on the retained predictions for the different uncertainty measures at three different uncertainty thresholds (0.01, 0.1 and 0.5).
Predictions below the uncertainty threshold are used to calculate performance metrics (TPR and FDR) for different thresholds on the prediction masks (i.e. sigmoid threshold used to binarize the network’s predictions is varied).
**Hypothesis:** If the predictions that are incorrect are uncertain, this filtering should increase the performance on remaining predictions.
TPR being defined by TP / (TP+FN). FDR by FP / (FP + TP).

Also Done:
- `dev/filtering_lesion.py`: compute percentage of voxels / lesion retained after uncertainty-filtering.

### CODE REVIEW

**Comment by charleygros** [comment]
TODO: Implement a normalisation process for lesion-uncertainty measures

It can be done by rescaling the lesion-uncertainties to [0,1] by subtracting by the minimum lesion uncertainty and dividing by the range.

Done for each measure separately.

Allow comparison across measures.

**Comment by charleygros** [comment]
`EvaluationMetric3D` has now data (numpy array) as inputs instead of fname, [here](https://github.com/neuropoly/ivado-medical-imaging/blob/master/ivadomed/utils.py#L46).
It allows to create such an object without having to save the numpy array as nifiti.

**Comment by charleygros** [comment]
Introducing a "config" dictionary to set-up the experiments:
```
    exp_dct = {
                'exp1': {'level': 'vox',
                            'uncertainty_measure': '_unc-vox',
                            'uncertainty_thr': [1e-5, 1e-3, 1e-1, 0.5],
                            'prediction_thr': [1e-6]+[t/10. for t in range(1,10,1)],
                            'results': {}},
                'exp2': {'level': 'obj',
                            'uncertainty_measure': '_unc-cv',
                            'uncertainty_thr': [1e-5, 1e-3, 1e-1, 0.5],
                            'prediction_thr': [1e-6]+[t/10. for t in range(1,10,1)],
                            'results': {}},
                'exp3': {'level': 'obj',
                            'uncertainty_measure': '_unc-avgUnc',
                            'uncertainty_thr': [1e-5, 1e-3, 1e-1, 0.5],
                            'prediction_thr': [1e-6]+[t/10. for t in range(1,10,1)],
                            'results': {}} 
    }
```

**Comment by charleygros** [comment]
`print_unc_stats` is a simple utils function that display stats measured on the Uncertainty maps (median, min etc.) in order to tailor the thresholds for each measure.

**Comment by charleygros** [comment]
Some experiments run on the SC T2star.

```
Mean percentage of retained elt:
        Unc threshold: 1e-05 --> 95.58718707318764
        Unc threshold: 0.001 --> 99.85090543731077
        Unc threshold: 0.1 --> 99.89460830551083
        Unc threshold: 0.5 --> 99.94736802986641
```
--> When the Uncertainty threshold increases, more (uncertain) voxels are retained.
The above percentages show the percentage of the non-zero voxels on the predictions that will not be set to zero by the uncertainty thresholding --> 95% means that 5% of the non-zero voxels are set to zero in the prediction, based on the uncertainty information.

AUC analyses: We fix the uncertainty threshold and we vary the threshold that binirise the predictions (`Pred Thr`):
```
Unc Thr: 1e-05
AUC: 0.6608487264416756, Optimal Pred Thr: 0.1
Unc Thr: 0.001
AUC: 0.7374577770228482, Optimal Pred Thr: 0.1
Unc Thr: 0.1
AUC: 0.7467468264010076, Optimal Pred Thr: 0.1
Unc Thr: 0.5
AUC: 0.7837295561953983, Optimal Pred Thr: 0.2
```
In this case, the optimal thresholds are: `Unc Thr = 0.5` and `Pred Thr = 0.2`.

File: ivadomed/main.py [hunk 1]
```diff
@@ -214,12 +214,20 @@ def cmd_train(context):
         for param in model.parameters():
             param.requires_grad = False
 
+        # TMP
+        model.decoder = models.Decoder(out_channel=context['out_channel'],
```
Comment (charleygros) [inline ivadomed/main.py:218]: https://github.com/neuropoly/ivado-medical-imaging/issues/189

**Comment by charleygros** [comment]
Off-topic new features:
1. Exploration: change the amount of layers that we freeze during the transfer learning. More to come: https://github.com/neuropoly/ivado-medical-imaging/issues/189
2. Fine tuning some part of the code after merging the multiclass PR #160 
3. Some "imports" cleaning

File: ivadomed/main.py (hunk 1 above)
Comment (olix86) [reply ivadomed/main.py:218]: So for now it retrains the entire decoder?

File: ivadomed/main.py [hunk 2]
```diff
@@ -625,7 +633,7 @@ def cmd_test(context):
     metric_mgr = metrics.MetricManager(metric_fns)
 
     # number of Monte Carlo simulation
-    if (context['uncertainty']['epistemic'] or context['uncertainty']['epistemic']) and \
```
Comment (olix86) [inline ivadomed/main.py:628]: 🙈

File: ivadomed/main.py [hunk 3]
```diff
@@ -811,10 +821,30 @@ def cmd_eval(context):
...
+        h, w, d = data_pred.shape[:3]
+        n_classes = len(fname_gt)
+        data_gt = np.zeros((h, w, d, n_classes))
+        for idx, file in enumerate(fname_gt):
+            if os.path.exists(file):
+                data_gt[..., idx] = self.get_data(file)
+            else:
+                data_gt[..., idx] = np.zeros((h, w, d), dtype='u1')
+
+        eval = Evaluation3DMetrics(data_pred=data_pred,
```
Comment (olix86) [inline ivadomed/main.py:836]: utils.Evaluation3DMetrics ?
